src/DisputeHandler.ts

```diff
@@ -55,164 +72,281 @@ class DisputeHandler {
-        } = this.agreementManager.getFinalizedAndLatestWithVotes(
+        const { dispute, disputeAuditingData } = await this.createDisputeStruct(
             forkCnt,
-            this.signerAddress
+            transactionCnt,
+            proofs,
+            timestamp,
+            timeout
         );
 
-        this.p2pEventHooks.onInitiatingDispute?.();
         await retry(
             async () => {
                 const txResponse =
                     await this.stateChannelManagerContract.createDispute(
-                        this.channelId,
-                        forkCnt,
-                        encodedLatestFinalizedState,
-                        encodedLatestCorrectState,
-                        virtualVotingBlocks,
-                        foldedParticipant,
-                        foldedTransactionCnt,
-                        proofs,
-                        { gasLimit: 4000000 } //TODO! - gas limit
+                        dispute
+                    );
+                const txReceipt = await txResponse.wait();
+                return txReceipt;
+            },
+            {
+                maxRetries: 1, // Current implementation retries once
+                onRetry: (attempt, error) => {
+                    console.log("ERROR - DISPUTE CATCH ##", error);
+                    console.log(
+                        `Retrying dispute creation, attempt ${attempt}`
+                    );
+                }
+            }
+        );
+
+        // gossip to all peers
+    }
+
+    public async auditDispute(
+        dispute: DisputeStruct,
+        disputeAuditingData: DisputeAuditingDataStruct
+    ): Promise<AddressLike[]> {
+        let slashParticipants: AddressLike[] = [];
+        await retry(
+            async () => {
+                slashParticipants =
+                    (await this.stateChannelManagerContract.auditDispute(
+                        dispute,
+                        disputeAuditingData
+                    )) as unknown as AddressLike[];
+                return slashParticipants;
+            },
+            {
+                maxRetries: 1, // Current implementation retries once
+                onRetry: (attempt, error) => {
+                    console.log("ERROR - DISPUTE CATCH ##", error);
+                    console.log(
+                        `Retrying dispute creation, attempt ${attempt}`
+                    );
+                }
+            }
+        );
+        return slashParticipants;
+    }
+
+    public async challengeDispute(
+        dispute: DisputeStruct,
+        newDispute: DisputeStruct,
+        disputeAuditingData: DisputeAuditingDataStruct
+    ): Promise<void> {
+        await retry(
+            async () => {
+                const txResponse =
+                    await this.stateChannelManagerContract.challengeDispute(
+                        dispute,
+                        newDispute,
+                        disputeAuditingData
```

Comment:
This is the interface that the on-chain contract expects - where do you construct the newDispute? It has to take into account the successful on-chain challenge